XenCert/StorageHandler.py

```diff
@@ -3170,3 +3169,105 @@ def ControlPathStressTests(self):
         XenCertPrint("Checkpoints: %d, totalCheckPoints: %s" % (checkPoint, totalCheckPoints))
 
         return (retVal, checkPoint, totalCheckPoints)
+
+    def DataIntegrityTests(self):
+        retVal = None
+        checkPoint = 0
+        totalCheckPoints = 11
+
+        vm_uuid = StorageHandlerUtil._get_localhost_uuid()
+        XenCertPrint("Got vm_uuid as %s" % vm_uuid)
+        vm_ref = self.session.xenapi.VM.get_by_uuid(vm_uuid)
+        sr_ref = None
+        vdi_ref = None
+        vbd_ref = None
+
+        try:
+            #1) Create SR
+            (retVal, sr_ref, dconf) = self.Create()
+            Print("Created SR")
+            checkPoint += 1
+
+            #2) Create 4GB VDI in SR
+            # XXX: Make Create_VDI() return vdi_ref on success and raise an
+            #      exception on failure. This is ugly and stoopid.
+            (retVal, vdi_ref) = self.Create_VDI(
+                    sr_ref,
+                    4 * StorageHandlerUtil.GiB)
+            if not retVal:
+                raise Exception('Error in VDI creation: %s' % vdi_ref)
+            Print("Created 4GB VDI")
+            checkPoint += 1
+
+            #3) Attach VDI to dom0
+            vbd_ref = StorageHandlerUtil.Attach_VDI(self.session, vdi_ref,
+                    vm_ref)
+            Print("Attached the VDI to dom0")
+            checkPoint += 1
+
+            #4) Write known pattern to VDI
+            StorageHandlerUtil.WriteDataToVDI(self.session, vbd_ref, 0, 3)
+            Print("Wrote data to VDI")
+            checkPoint += 1
+
+            #5) Detach VDI
+            StorageHandlerUtil.Detach_VDI(self.session, vbd_ref)
+            Print("Detached from dom0")
+            checkPoint += 1
+
+            #6) Resize VDI to 8GB
+            self.Resize_VDI(vdi_ref, 8 * StorageHandlerUtil.GiB)
+            Print("Resized the VDI to 8GB")
+            checkPoint += 1
+
+            #7) Attach VDI to dom0
+            vbd_ref = StorageHandlerUtil.Attach_VDI(self.session, vdi_ref,
+                    vm_ref)
+            Print("VDI attached again to Dom0")
+            checkPoint += 1
+
+            #8) Write known pattern to second 4GB chunk
+            StorageHandlerUtil.WriteDataToVDI(self.session, vbd_ref, 4, 7)
```

Comment:
I know we didn't introduce WriteDataToVDI as part of this changeset but we are using it. We don't close any open file descriptors when an exception occurs in the method which can cause the subsequent Detach_VDI to hang. Also true for VerifyDataOnVDI